### ISSUE TITLE
Add integration tests for timeouts and retries

### ISSUE THREAD

**Issue by adleong** [linkerd/linkerd2#2179]
We should add integration tests that exercise timeouts and retries individually any together.  

Some behaviors to verify include:
* Failures on retryable routes are retried
* Retries do not exceed retry budget
* Route request actual stats reflect actual requests sent
* Route request stats reflect logical requests
* Long requests are timed out when a timeout is sent
* Server receives a cancellation when a request times out
* Client receives a 504 when request times out
* Retries stop after the timeout is reached
* Timed out responses have `err=timeout` label

**Comment by liquidslr** [comment]
Hy @adleong I would like to work on this issue.

**Comment by adleong** [comment]
@liquidslr awesome!  To get started, I'd recommend taking a look at https://github.com/linkerd/linkerd2/blob/master/TEST.md#writing-tests and reading some of the existing integration tests in the `/test` directory.  Please let me know if you need any further guidance!

**Comment by liquidslr** [comment]
thanks for the update @adleong, just needed a small help in getting started, can you briefly explain the requirements of these tests and related information and when will these tests run?

**Comment by adleong** [comment]
@liquidslr: The requirements would just be to verify the behaviors listed in the issue.

Unfortunately, this test is a bit different from our existing integration tests because it needs to verify the behavior of individual requests rather than just passively observing existing traffic.

There are probably a number of different ways to accomplish this but here's one suggestion: 
Create a new test application similar to [bb](https://github.com/BuoyantIO/bb) that accepts requests with three parameters: `id`, `succeed-after-tries`, and `delay`.  The service keeps a map of `id` to `tries` which keeps track of how many times the service has been called for that id.  If the number of tries is greater than the value of `succeed-after-tries` then it returns a 200 OK.  Otherwise, it returns a 500 Error.  This ways we can issue requests to the service which will fail, get retried, and eventually succeed.  The service should also sleep for `delay` before responding so that we can test the timeout logic as well.

You'll also need a gateway service, injected with Linkerd, which will call your test service.  You can probably use `bb` as the gateway.

The body of the tests would be to deploy the test service and gateway, injected with Linkerd.  Next it would set up a port-forward to the gateway.  Finally, the test would manually issue requests to the gateway through the port-forward.  There are a number of helpers in `test_helper.go` that can help with much of this.

As I said, there are probably other ways to test this as well.  @klingerf, do you have any other suggestions for ways to set up tests like this?

**Comment by klingerf** [comment]
Yep, 👍 to @adleong's suggestion. I actually recommend making the succeed-after-tries change that he mentions directly in bb itself, so that we don't end up with a proliferation of test apps. And bb already supports a `--sleep-in-millis` flag, which you can use for testing timeout logic.

**Comment by liquidslr** [comment]
> Create a new test application similar to [bb](https://github.com/BuoyantIO/bb)

@adleong should I create the test application inside the test directory ?

**Comment by kohsheen1234** [comment]
@adleong I would like to take up this issue :)

**Comment by ChinYing-Li** [comment]
Hi, I am wondering if this issue is still relevant? If so, I'd like to tackle this one.

**Comment by adleong** [comment]
Hi @ChinYing-Li!  This issue is still relevant, please ask if you have any questions!
